dowhy/gcm/validation.py

```diff
@@ -0,0 +1,157 @@
+        if parents and non_descendants:
+            # test local Markov condition, null=conditional independence
+            lmc_p_value = conditional_independence_test(data[node].values,
+                                                        data[non_descendants].values,
+                                                        data[parents].values)
+            lmc_test_result = dict(p_value=lmc_p_value)
+            all_p_values.append(lmc_p_value)
+
+        # test edge dependence, null=independence
+        edge_dependence_result = dict()
+        for parent in parents:
+            edge_p_value = independence_test(data[parent].values, data[node].values)
+            edge_dependence_result[parent] = dict(p_value=edge_p_value)
+            all_p_values.append(edge_p_value)
+
+        validation_summary[node] = dict(
+            local_markov_test=lmc_test_result,
+            edge_dependence_test=edge_dependence_result
+        )
+
+    if fdr_control_method is None:
+        successes = np.array(all_p_values) <= significance_level
+        adjusted_p_values = np.empty(len(successes))
+        adjusted_p_values[:] = np.nan
+    else:
+        multipletests_result = multipletests(all_p_values, significance_level, method=fdr_control_method)
+        successes = multipletests_result[0]
+        adjusted_p_values = multipletests_result[1]
+
+    # The order of the p-values added to the list is deterministic.
+    index = 0
+    for node in causal_graph.nodes:
+        if 'p_value' in validation_summary[node]['local_markov_test']:
+            validation_summary[node]['local_markov_test']['fdr_adjusted_p_value'] \
+                = adjusted_p_values[index]
+            validation_summary[node]['local_markov_test']['success'] \
+                = not successes[index]
+            is_dag_valid &= not successes[index]
+            index += 1
+
+        for parent in get_ordered_predecessors(causal_graph, node):
+            validation_summary[node]['edge_dependence_test'][parent]['fdr_adjusted_p_value'] \
+                = adjusted_p_values[index]
+            validation_summary[node]['edge_dependence_test'][parent]['success'] \
+                = successes[index]
+            is_dag_valid &= successes[index]
+            index += 1
+    if is_dag_valid:
+        return RejectionResult.NOT_REJECTED, validation_summary
+    else:
+        return RejectionResult.REJECTED, validation_summary
+
+
+def validate_pnl_assumptions(causal_model: InvertibleStructuralCausalModel,
+                             data: pd.DataFrame,
+                             independence_test: Callable[[np.ndarray, np.ndarray], float] = kernel_based,
+                             significance_level: float = 0.05,
+                             fdr_control_method: Optional[str] = None) -> bool:
+    """Validate the assumption that the structural causal models can be represented by :py:class:`PostNonlinearSEM <causality.sem.PostNonlinearSEM>` (e.g.
```

Comment:
If the test fails, it can be due to two reasons, 1) the structure is incorrect, 2) functions that were fitted were not ideal (e.g., linear regression instead of non-linear model). 
Is there a way to know which of the two is the reason?

Reply:
right, this makes sense. thanks for the explanation.